Kconfig.headers

```diff
@@ -0,0 +1,36 @@
+# Configurations for headers
+# This file can't be added like other Kconfig files since putting it under include/ will make
+# it visibile to IDEs when writing 'include' statements.
+
+# Copyright (c) 2022 Google Inc
+# SPDX-License-Identifier: Apache-2.0
+
+config MATH_UTIL_FLOAT
+	bool "Enable the math utility library to use floating point values"
+	depends on FPU
+	help
+	  When enabled, 'fp_t' type will be mapped to 'float'. Otherwise, it will use int32_t and
+	  represent fixed point math.
+
+config MATH_UTIL_FP_BITS
```

Comment:
Similar to the above comment really, I don't believe a Kconfig option should be drastically changing the way in which project wide math works.

If varying q formats are useful, then perhaps make those as new types. If you perhaps assume _Generic exists it makes it maybe even easier to write a common fp_div/fp_mul and perhaps some saturating operations which are useful. These could wrap format specific versions.

I don't really think we need to enable every fixed point format here. A few useful ones would be fine. e.g. sensor_value could be replaced by typedef uint64_t q31_32_t easily and effectively fit all the known sensors in the tree. For normalized values q1.15, q1.31 are likely *plenty* as shown by arm's cmsis-dsp.

e.g. typedef uint32_t q15_16_t;

So maybe again, a short list of fixed point formats, matching format strings, mul/div, and conversion functions to float, integers, and maybe integer with a divider.

An example maybe helps clarify my thinking here...

I have an int representing 3 microsecond, I want to convert it to fixed point with seconds as the measurement unit.

Ideally any conversion should validate the value is at least somewhat representable.

So perhaps mn_to_q15_16(3, 1000000); should fail on an assert.

Something like mn_to_q31_32(3, 1000000); or mn_to_q1_31(3, 1000000); should be ok.

What I don't want is this which bit me once already...

fp_t seconds = fp_div(INT_TO_FP(3), INT_TO_FP(1000000));

What should not happen, in my opinion, is the above sort of conversion failing or succeeding based on a project wide Kconfig.

Reply:
I can easily do this in C++ using a template but I can't think of a good way to do this in C.  Open to suggestions. I worry about typedef as it can still be mixed and cause errors.

Reply:
Yeah, I'm not finding a good way to do this in C, in C++ it's easy:

```
constexpr fp_t FLOAT_TO_FP(float x) {
  fp_inter_t intermediate = (fp_inter_t)(x * (float)(1 << 16));
  assert(intermediate < FLT_MAX);
  assert(intermediate > FLT_MIN);
  return (fp_t)intermediate;
}
```

Reply:
For making sure that different values aren't combined with different resolution:

```
template <int decimal_bits>
struct fp {
  union {
    int32_t raw;
    struct {
      int32_t integer : 32 - decimal_bits;
      int32_t decimal : decimal_bits;
    };
  };
};
```

Now you can do:
```
typedef struct fp<16> q15_16;
typedef struct fp<8> q23_8;
...
```

Reply:
You can do the same sort of thing with a macro in C, its not as good as a const expr, but a statement expr macro would be close enough


``` c
/** Define a q1_31_t */
typedef struct { int32_t q1_31; } q1_31_t;

/**
 * Convert a value of a numerical rval to a qval
 * and validate it fits in the bounds of the qval
 * close equivalent to q = val; with bounds checking.
 */
#define q_convert(qval, rval) \
do { \
  int32_t whole = z_q_whole(qval, rval); \
  int32_t fraction = z_q_fract(qval, rval); \
  z_q_assign(qval, whole, fract); \
} while (false)
   
```
 
So something like the above perhaps might work out, where you *know* the type of q being assigned and the type you are converting from.

It requires use of _Generic((qval), q1_31_t: ..., ) style macros which... should be fine.

It avoids operator mismatching, everything must be done with macros

q_add, q_sub, q_mul, q_div, q_mod, q_rem, q_abs, etc, etc, etc

And many of them likely require type specialization which can be done with _Generic.

The fewer q types you have, the saner this ends up.

I know its not as sane as say templates, but its possible I believe.

Reply:
`_Generic` is neat, I haven't seen it before. I'll see if I can get this to work.

Reply:
Note that `_Generic` is a C11 feature.

Reply:
It wouldn't be the first usage of it in the tree, though it would push us to say _Generic is whats required for this. I don't think that's necessarily a bad thing.

Reply:
Maybe one more nifty feature of this... and probably should be a requirement.

If the hardware supports fixed point instructions of certain q types and operations, these could should easily be overridden with those.

Reply:
C11 is already documented as "highly recommended" https://docs.zephyrproject.org/latest/develop/languages/c/index.html?highlight=c11#c-language-support